ompi/mpiext/continue/c/continuation.c

```diff
@@ -0,0 +1,671 @@
+        /* global progress */
+        while (max > completed) {
+            ompi_continuation_t *cb;
+            if (using_threads) {
+                opal_mutex_lock(&request_cont_lock);
+                cb = (ompi_continuation_t*)opal_list_remove_first(&continuation_list);
+                opal_mutex_unlock(&request_cont_lock);
+            } else {
+                cb = (ompi_continuation_t*)opal_list_remove_first(&continuation_list);
+            }
+            if (NULL == cb) break;
+            ompi_continue_cont_invoke(cb);
+            ++completed;
+        }
+    }
+
+    in_progress = 0;
+
+    return completed;
+}
+
+static int ompi_continue_progress_callback()
+{
+    return ompi_continue_progress_n(1);
+}
+
+int ompi_continue_progress_request(ompi_request_t *req)
+{
+    if (in_progress) return 0;
+    ompi_cont_request_t *cont_req = (ompi_cont_request_t *)req;
+    if (NULL == cont_req->cont_complete_list) {
+        /* progress as many as allowed */
+        return ompi_continue_progress_n(cont_req->continue_max_poll);
+    }
+    if (opal_list_is_empty(cont_req->cont_complete_list)) {
+        return 0;
+    }
+
+    in_progress = 1;
+
+    const uint32_t max_poll = cont_req->continue_max_poll;
+
+    uint32_t completed = 0;
+    const bool using_threads = opal_using_threads();
+    while (max_poll > completed && !opal_list_is_empty(cont_req->cont_complete_list)) {
+        ompi_continuation_t *cb;
+        if (using_threads) {
+            opal_atomic_lock(&cont_req->cont_lock);
+            cb = (ompi_continuation_t *) opal_list_remove_first(cont_req->cont_complete_list);
+            opal_atomic_unlock(&cont_req->cont_lock);
+        } else {
+            cb = (ompi_continuation_t *) opal_list_remove_first(cont_req->cont_complete_list);
+        }
+        if (NULL == cb) break;
+
+        ompi_continue_cont_invoke(cb);
+        completed++;
+    }
+
+    in_progress = 0;
+
+    return completed;
+}
+
+
+/**
+ * Register the provided continuation request to be included in the
+ * global progress loop (used while a thread is waiting for the contnuation
+ * request to complete).
+ * We move all local continuations into the global continuation list
+ * and mark the continuation request such that future continuations
+ * are directly put into the global continuations list.
+ * Once the wait completed (i.e., all continuations registered with the
+ * continuation request) we unmark it (see ompi_continue_deregister_request_progress).
+ */
+int ompi_continue_register_request_progress(ompi_request_t *req)
+{
+    ompi_cont_request_t *cont_req = (ompi_cont_request_t *)req;
+
+    if (NULL == cont_req->cont_complete_list) return OMPI_SUCCESS;
+
+    opal_atomic_lock(&cont_req->cont_lock);
+
+    cont_req->cont_in_wait = true;
+
+    ompi_continue_cont_req_release(cont_req, opal_list_get_size(cont_req->cont_complete_list), false);
+
+    opal_atomic_unlock(&cont_req->cont_lock);
+
+    if (NULL == thread_progress_list) {
```

Comment:
Why is this global object not protected for concurrent accesses ?

Reply:
That's a thread_local variable that stores the continuation requests the current thread is waiting on